tests/unit/steps/test_steps_plan.py

```diff
@@ -638,4 +582,219 @@ def test_verify_data_plane_cli_azs_raise_cannot_find():
+
+    hypervisors = await cou_plan.filter_hypervisors_machines(cli_args, analysis_result)
+
+    assert hypervisors == expected_hypervisors
+
+
+@pytest.mark.parametrize(
+    "force, novas_machines_input, novas_machines_expected",
+    [
+        (False, [0, 1, 2], [0, 1]),
+        (False, [2], []),
+        (False, [0, 1], [0, 1]),
+        (False, [0], [0]),
+        (True, [0, 1, 2], [0, 1, 2]),
+        (True, [2], [2]),
+        (True, [0], [0]),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.steps.plan._get_hypervisors_machines_possible_to_upgrade")
+async def test_filter_hypervisors_machines_by_az(
+    mock_hypervisors_machines,
+    force,
+    novas_machines_input,
+    novas_machines_expected,
+    apps_machines,
+    cli_args,
+    analysis_result,
+):
+    expected_hypervisors, hypervisors_to_upgrade = _generate_expected_hypervisors_to_test(
+        force, novas_machines_expected, apps_machines
+    )
+    mock_hypervisors_machines.return_value = hypervisors_to_upgrade
+
+    cli_azs = {
+        apps_machines["nova-compute"][NOVA_MACHINES[machine_id]].az
+        for machine_id in novas_machines_input
+    }
+
+    cli_args = _prepare_cli_args(cli_args, None, None, cli_azs, force)
+
+    hypervisors = await cou_plan.filter_hypervisors_machines(cli_args, analysis_result)
+
+    assert hypervisors == expected_hypervisors
+
+
+@pytest.mark.parametrize(
+    "force, novas_machines_expected",
+    [
+        (False, [0, 1]),
+        (True, [0, 1, 2]),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.steps.plan._get_hypervisors_machines_possible_to_upgrade")
+async def test_filter_hypervisors_machines_no_parameter(
+    mock_hypervisors_machines,
+    force,
+    novas_machines_expected,
+    apps_machines,
+    cli_args,
+    analysis_result,
+):
+    expected_hypervisors, hypervisors_to_upgrade = _generate_expected_hypervisors_to_test(
+        force, novas_machines_expected, apps_machines
+    )
+    mock_hypervisors_machines.return_value = hypervisors_to_upgrade
+
+    cli_args = _prepare_cli_args(cli_args, None, None, None, force)
+
+    hypervisors = await cou_plan.filter_hypervisors_machines(cli_args, analysis_result)
+
+    assert hypervisors == expected_hypervisors
+
+
+def _generate_expected_hypervisors_to_test(force, novas_machines_expected, apps_machines):
+    """Wrap the generation of the expected hypervisors to upgrade for testing."""
+    expected_hypervisors = {
+        apps_machines["nova-compute"][NOVA_MACHINES[machine_id]]
+        for machine_id in novas_machines_expected
+    }
+
+    # assume that machine 2 is running a vm.
+    hypervisors_to_upgrade = {
+        apps_machines["nova-compute"][NOVA_MACHINES[0]],
+        apps_machines["nova-compute"][NOVA_MACHINES[1]],
+    }
+    if force:
+        hypervisors_to_upgrade.add(apps_machines["nova-compute"][NOVA_MACHINES[2]])
+
+    return expected_hypervisors, hypervisors_to_upgrade
+
+
+def _prepare_cli_args(cli_args, machines, hostnames, az, force):
```

Comment:
This is not good behavior to change input argument, we should avoid usage of these and especially in tests, where it must be obvious what is going on.